programs/viewer/PluginLoader.js

```diff
@@ -24,49 +24,173 @@
+                getClass: function() { return ODFViewerPlugin; }
+            };
+        }())
+    ];
+
+    function loadPlugin(pluginName, callback) {
+        "use strict";
+        var script, style;
+
+        // Load script
+        script = document.createElement('script');
+        script.async = false;
+        script.onload = callback;
+        script.src = pluginName + '.js';
+        script.type = 'text/javascript';
+        document.getElementsByTagName('head')[0].appendChild(script);
+    }
+
+
+    function estimateTypeByHeaderContentType(documentUrl, cb) {
+        var xhr = new XMLHttpRequest();
+        xhr.onreadystatechange = function() {
+            var mimetype, matchingPluginData;
+            if (xhr.readyState === 4) {
+                if ((xhr.status >= 200 && xhr.status < 300) || xhr.status === 0) {
+                    mimetype = xhr.getResponseHeader('content-type');
+
+                    if (mimetype) {
+                        pluginRegistry.some(function(pluginData) {
+                            if (pluginData.supportsMimetype(mimetype)) {
+                                matchingPluginData = pluginData;
+                                console.log('Found plugin by mimetype and xhr head: ' + mimetype);
+                                return true;
+                            }
+                            return false;
+                        });
+                    }
+                }
+                cb(matchingPluginData);
+            }
+        };
+        xhr.open("HEAD", documentUrl, true);
+        xhr.send();
+    }
+
+
+    function doEstimateTypeByFileExtension(extension) {
+        var matchingPluginData;
+
+        pluginRegistry.some(function(pluginData) {
+            if (pluginData.supportsFileExtension(extension)) {
+                matchingPluginData = pluginData;
+                return true;
+            }
+            return false;
+        });
+
+        return matchingPluginData;
+    }
+
+
+    function estimateTypeByFileExtension(extension) {
+        var matchingPluginData = doEstimateTypeByFileExtension(extension)
+
+        if (matchingPluginData) {
+            console.log('Found plugin by parameter type: ' + extension);
         }
+
+        return matchingPluginData;
+    }
+
+
+    function estimateTypeByFileExtensionFromPath(documentUrl) {
+        // See to get any path from the url and grep what could be a file extension
+        var documentPath = documentUrl.split('?')[0],
+            extension = documentPath.split('.').pop(),
+            matchingPluginData = doEstimateTypeByFileExtension(extension)
+
+        if (matchingPluginData) {
+            console.log('Found plugin by file extension from path: ' + extension);
+        }
+
+        return matchingPluginData;
+    }
+
+
+    function parseSearchParameters(location) {
+        var parameters = {},
+            search = location.search || "?";
+
+        search.substr(1).split('&').forEach(function (q) {
```

Comment:
Hm, how does the splitting logic fail here exactly?

Actually it might be the webserver failing here. Because at least the `python -m SimpleHTTPServer` one does not get `/programs/viewer/?title=T` properly, seems to assume a full path for which there is no match and redirects to `/programs/viewer/?title=T\`, for which then suddenly `programs/viewer/index.html` is served:
```
127.0.0.1 - - [29/Jan/2015 02:51:40] "GET /programs/viewer/?title=T HTTP/1.1" 301 -
127.0.0.1 - - [29/Jan/2015 02:51:40] "GET /programs/viewer/?title=T/ HTTP/1.1" 200 -
```
Try with `node webodf/httpserver.js` and see it behave better :)

The only issue here I see is that empty string by `&&` or those with no `=` are implicitely dealt with, the first by adding a property with empty string and the second by setting a null value for a key that is the full string. Which is both fine IMHO, but could be documented.
Added skipping of empty strings and added a note for handling of no `=`.